File: hw1/particles.py
import itertools
import numpy

class Particle(object):
    def __init__(self, mass, radius, xPos, yPos, xVel, yVel, maxX, maxY):
        """
           mass in g
           radius in cm
           xPos, yPos cm
           xVel, yVel cm/s
        """
        self.mass = mass
        self.radius = radius
        self.xPos = xPos
        self.yPos = yPos
        self.xVel = xVel
        self.yVel = yVel
        self.maxX = maxX
        self.maxY = maxY

    def updatePosition(self, dt):
        """Propagate particle according to velocities
        reverse velocity if we're at max (wall boundary)
        """
        self.xPos = self.xPos + self.xVel * dt
        self.yPos = self.yPos + self.yVel * dt

        # reverse direction if we're at a wall
        if self.xPos + self.radius > self.maxX and self.xVel > 0:
            self.xVel = -1 * self.xVel
        if self.xPos - self.radius < 0 and self.xVel < 0:
            self.xVel = -1 * self.xVel
        if self.yPos + self.radius > self.maxY and self.yVel > 0:
            self.yVel = -1 * self.yVel
        if self.yPos - self.radius < 0 and self.yVel < 0:
            self.yVel = -1 * self.yVel

    @property
    def speed(self):
        """Speed of particle"""
        return numpy.sqrt(self.xVel**2 + self.yVel**2)

    @property
    def p(self):
        """momentum of particle in x and y"""
        return self.mass * self.xVel, self.mass * self.yVel

    @property
    def ke(self):
        """kinetic energy of particle"""
        return 0.5 * self.mass * (self.xVel**2 + self.yVel**2)

    @property
    def motionAngle(self):
        """Return the direction of motion as an angle
        """
        return numpy.arctan2(self.yVel, self.xVel)


class Box(object):
    def __init__(self, width, height):
        """input width / height in cm
        """
        # 0,0 is box center
        self.width = width
        self.height = height
        self.particles = []
        self.dt = None # set at simulation time

    def randomPoint(self, radius):
        """Return a random xy point inside of box
        """
        x = numpy.random.uniform(0+radius, self.width-radius)
        y = numpy.random.uniform(0+radius, self.height-radius)
        return x, y

    def isCollided(self, particle1, particle2):
        """Return True if these particles are collided
        """
        dist = numpy.linalg.norm(
            [particle1[2] - particle2[2], particle1[3] - particle2[3]]
        )
        return dist < (particle1[1] + particle2[1])


    def findCollision(self, pIndex):
        """Check if this particle is collided, expect that is
        may only be collided with 1 other particle

        return the set {pIndex, collidedNeighbor}
        """

        # calculate distance to all particles
        xys = self.particles[:,2:4]
        radii = self.particles[:,1]
        dxys = xys - xys[pIndex]
        dist = numpy.linalg.norm(dxys,axis=1)
        # note we can't handle differing radii
        collidedIndices = set(numpy.argwhere(dist <= 2*radii).flatten())
        nCollided = len(collidedIndices)
        if nCollided > 2:
            print("warning: > 2 body collision, shouldn't happen: %i"%nCollided)
        if nCollided == 1:
            # ignore a self-collision!
            return None
        # returns, {pIndex, collidedIndex}
        return set(collidedIndices)


    def handleCollision(self, p1index, p2index):
        """Modify the velocities of particle 1 and 2 for an elastic collision

        arguments are indices in particle list

        https://en.wikipedia.org/wiki/Elastic_collision
        """


        # contact angle
        ca = numpy.arctan2(
            self.particles[p1index][3] - self.particles[p2index][3],
            self.particles[p1index][2] - self.particles[p2index][2]
        )

        # calculate motion angles for particles
        ma1 = numpy.arctan2(self.particles[p1index][5], self.particles[p1index][4])
        ma2 = numpy.arctan2(self.particles[p2index][5], self.particles[p2index][4])

        # calculate speed for particles
        sp1 = numpy.linalg.norm(self.particles[p1index][4:6])
        sp2 = numpy.linalg.norm(self.particles[p2index][4:6])

        # masses of particles
        m1 = self.particles[p1index][0]
        m2 = self.particles[p2index][0]


        A1 = (
            sp1 * numpy.cos(ma1 - ca) * (m1 - m2) +
            2 * m2 * sp2 * numpy.cos(ma2 - ca)
        ) / (m1 + m2)

        A2 = (
            sp2 * numpy.cos(ma2 - ca) * (m2 - m1) +
            2 * m1 * sp1 * numpy.cos(ma1 - ca)
        ) / (m2 + m1)

        # new x velocity for particle 1
        v1x = A1 * numpy.cos(ca) + sp1 * numpy.sin(ma1 - ca) * numpy.cos(ca + numpy.pi / 2)
        # new y velocity for particle 1
        v1y = A1 * numpy.sin(ca) + sp1 * numpy.sin(ma1 - ca) * numpy.sin(ca + numpy.pi / 2)

        # new x velocity for particle 2
        v2x = A2 * numpy.cos(ca) + sp2 * numpy.sin(ma2 - ca) * numpy.cos(ca + numpy.pi / 2)
        # new y velocity for particle 2
        v2y = A2 * numpy.sin(ca) + sp2 * numpy.sin(ma2 - ca) * numpy.sin(ca + numpy.pi / 2)

        self.particles[p1index][4] = v1x
        self.particles[p1index][5] = v1y
        self.particles[p2index][4] = v2x
        self.particles[p2index][5] = v2y

    def addRandomParticle(self, mass, radius, ke):
        """Add a random particle to the box, give it a random position
        and random direction with kinetic energy specified
        id : an identifier for this particle
        mass: mass of particle in grams
        radius: radius of particle in cm
        ke: kinetic energy in gcm2/s2 (erg)
        """
        speed = numpy.sqrt(2 * ke / mass)
        # pick a random angle 0 - 2pi
        ang = numpy.random.uniform(0, 2*numpy.pi)
        xVel = numpy.cos(ang) * speed
        yVel = numpy.sin(ang) * speed
        # find a non-colliding place for it in the box
        while True:
            x, y = self.randomPoint(radius)
            p = [mass, radius, x, y, xVel, yVel]
            if not self.particles:
                print('added first particle')
                # this is first particle
                self.particles.append(p)
                break # from while loop

            # check if this particle collides with any others
            isCollided = False
            for otherParticle in self.particles:
                if self.isCollided(p, otherParticle):
                    isCollided = True
                    break

            if not isCollided:
                # particle isn't collided, save it
                self.particles.append(p)
                print('added %i particle'%(len(self.particles)))
                break # from while loop


    def updatePositions(self):
        """Propagate all particles according to their velocities,
        dt is specified in runSim()
        """
        self.particles[:, 2] = self.particles[:, 2] + self.particles[:, 4] * self.dt
        self.particles[:, 3] = self.particles[:, 3] + self.particles[:, 5] * self.dt

        # find particles at box edges and reverse their velocities
        # may want to check that velocities haven't been updated already?
        rightEdge = self.particles[:, 2] + self.particles[:, 1] >= self.width
        leftEdge = self.particles[:, 2] - self.particles[:, 1] <= 0
        topEdge = self.particles[:, 3] + self.particles[:, 1] >= self.height
        bottomEdge = self.particles[:, 3] - self.particles[:, 1] <= 0

        # modify x velocities for left and right edges
        self.particles[rightEdge, 4] = -1*self.particles[rightEdge, 4]
        self.particles[leftEdge, 4] = -1*self.particles[leftEdge, 4]

        # modify y velocities for left and right edges
        self.particles[topEdge, 5] = -1*self.particles[topEdge, 5]
        self.particles[bottomEdge, 5] = -1*self.particles[bottomEdge, 5]


    def runSim(self, maxSteps=10000, saveEvery=1, dt=None):
        """Start the simulation, pick a reasonable timestep based on ke
        and smallest particle radius
        """
        #
        self.simSteps = [] # holds snapshots of the simulation
        self.collisionFlags = []

        # convert particle list to numpy array for quicker computations
        self.particles = numpy.array(self.particles)

        if dt is None:
            # set the time step such that dt*maxSpeed < 0.01 the particle radius
            maxSpeed = numpy.max(numpy.linalg.norm(self.particles[:,4:6], axis=1))
            minRadius = numpy.min(self.particles[:,1])
            self.dt = 0.0005 * minRadius / maxSpeed
        else:
            self.dt = dt

        for step in range(maxSteps):
            print("step", step)
            # update positions propagates particles
            # it also reflects velocities at edges of box
            self.updatePositions()


            collidedPairs = []
            for ii in range(len(self.particles)):
                # findCollision returns a set
                collidedPair = self.findCollision(ii)
                if collidedPair is not None:
                    collidedPairs.append(collidedPair)

            # only operate on unique collisions
            # (we find both A-B and B-A collisions)
            collisionFlags = []
            for collidedPair in numpy.unique(collidedPairs):
                index1, index2 = list(collidedPair)
                self.handleCollision(index1, index2)
                collisionFlags.append(index1)
                collisionFlags.append(index2)


            # save state
            if step % saveEvery == 0:
                self.simSteps.append(numpy.copy(self.particles)) # ensures a copy
                self.collisionFlags.append(collisionFlags)

File: hw1/particles.cpp
#include <iostream>
#include <cmath>
#include "particles.h"

double randomSample(){
    // return between 0 and 1
    return static_cast <double> (rand()) / static_cast <double> (RAND_MAX);
}


Particle::Particle(double mass, double radius, double x, double y, double vx, double vy)
    : mass(mass), radius(radius), x(x), y(y), vx(vx), vy(vy){};


std::array<double, 2> Particle::nextPosition(double dt){
    std::array<double, 2> nextXY;
    nextXY[0] = x + vx*dt;
    nextXY[1] = y + vy*dt;
    return nextXY;

}

void Particle::updatePosition(double dt){
    auto nextXY = nextPosition(dt);
    x = nextXY[0];
    y = nextXY[1];
}

void Particle::setVelocities(double vxnext, double vynext){
    vx = vxnext;
    vy = vynext;
}

double Particle::speed(){
    return hypot(vx, vy);
}

double Particle::direction(){
    return atan2(vy, vx);
}


Box::Box(double width, double height, int seed)
    : width(width), height(height)
{
    srand(seed);
}

std::array<double, 2> Box::randomPoint(double radius){
    std::array<double, 2> xy;
    xy[0] = randomSample()*(width - 2*radius) + radius;
    xy[1] = randomSample()*(height - 2*radius) + radius;
    return xy;
}

bool Box::isCollided(std::shared_ptr<Particle> p1, std::shared_ptr<Particle> p2){
    double dx, dy, dist, nextDist;
    std::array<double, 2> p1next, p2next;

    dx = p1->x - p2->x;
    dy = p1->y - p2->y;
    dist = hypot(dx, dy); // current distance

    // first check if particles are moving towards eachother
    // if not, don't consider them collided
    p1next = p1->nextPosition(dt);
    p2next = p2->nextPosition(dt);
    nextDist = hypot(p1next[0]-p2next[0], p1next[1]-p2next[1]);
    if (nextDist > dist){
        // particles are moving away (collision probably already happened)
        return false;
    }

    if (dist <= p1->radius + p2->radius){
        // distance is within the collision zone
        return true;
    }

    return false;
}

void Box::addRandomParticle(double mass, double radius, double speed){
    bool collided;
    double ang, vx, vy;

    ang = randomSample()*2*M_PI;
    vx = cos(ang)*speed;
    vy = sin(ang)*speed;

    if (particles.empty()){
        // this is the first particle added
        // add it and exit
        auto xy = randomPoint(radius);
        auto p = std::make_shared<Particle>(mass, radius, xy[0], xy[1], vx, vy);
        particles.push_back(p);
        return;
    }

    // ensure we find a non-colliding position for the particle
    // try 100000 times before giving up and throwing a runtime error
    int maxIter = 100000;
    for (int ii=0; ii<=maxIter; ii++){
        collided = false;
        auto xy = randomPoint(radius);
        auto p = std::make_shared<Particle>(mass, radius, xy[0], xy[1], vx, vy);
        for (auto p2 : particles){
            if (isCollided(p, p2)){
                collided = true;
                break;
            }
        }

        if (!collided){
            particles.push_back(p);
            std::cout << "particles " << particles.size() << std::endl;
            break;
        }
        if (ii==maxIter){
            throw std::runtime_error("cannot find place for new particle");
        }
    }
}

void Box::handleCollision(std::shared_ptr<Particle> p1, std::shared_ptr<Particle> p2){
    // assumes collision has been verified for input particle indices...
    // implement the equations from:
    // https://en.wikipedia.org/wiki/Elastic_collision
    double contactAngle, A1, A2, v1x, v1y, v2x, v2y;

    contactAngle = atan2(
        p1->y - p2->y,
        p1->x - p2->x
    );

    A1 = (
        p1->speed() *
        cos(p1->direction() - contactAngle) *
        (p1->mass - p2->mass) +
        2 * p2->mass * p2->speed() *
        cos(p2->direction() - contactAngle)
    ) / (p1->mass + p2->mass);

    A2 = (
        p2->speed() *
        cos(p2->direction() - contactAngle) *
        (p2->mass - p1->mass) +
        2 * p1->mass * p1->speed() *
        cos(p1->direction() - contactAngle)
    ) / (p2->mass + p1->mass);

    // new x velocity for particle 1
    v1x = A1 * cos(contactAngle) +
        p1->speed() * sin(p1->direction() - contactAngle) *
        cos(contactAngle + M_PI / 2);
    // new y velocity for particle 1
    v1y = A1 * sin(contactAngle) + \
        p1->speed() * sin(p1->direction() - contactAngle) *
        sin(contactAngle + M_PI / 2);

    // new x velocity for particle 2
    v2x = A2 * cos(contactAngle) + \
        p2->speed() * sin(p2->direction() - contactAngle) *
        cos(contactAngle + M_PI / 2);
    // new y velocity for particle 2
    v2y = A2 * sin(contactAngle) + \
        p2->speed() * sin(p2->direction() - contactAngle) *
        sin(contactAngle + M_PI / 2);

    p1->setVelocities(v1x, v1y);
    p2->setVelocities(v2x, v2y);

}

std::vector<std::array<double, 5>> Box::dumpState(){
    std::vector<std::array<double, 5>> ensemble;

    for (auto p : particles){
        std::array<double, 5> pState;
        pState[0] = p->mass;
        pState[1] = p->radius;
        pState[2] = p->x;
        pState[3] = p->y;
        pState[4] = p->speed();
        ensemble.push_back(pState);
    }
    return ensemble;
}

void Box::runSim(int steps, double timestep, int saveEvery){
    double pressure, collisions;

    dt = timestep;
    int nParticles = particles.size();

    for (int step=0; step < steps; step++){
        // begin main sim loop
        // first propatate particles
        pressure = 0;
        collisions = 0;

        for (auto p : particles){
            p->updatePosition(dt);
            // reflect the particle if it's at the boundary
            if ( (p->x + p->radius >= width) and (p->vx > 0)){
                pressure += p->mass * 2 * p->vx / height;
                p->setVelocities(-1*p->vx, p->vy);
            }
            if ( (p->y + p->radius >= height) and (p->vy > 0)){
                p->setVelocities(p->vx, -1*p->vy);
                pressure += p->mass * 2 * p->vy / width;
            }
            if ( (p->x - p->radius <= 0) and (p->vx < 0)){
                p->setVelocities(-1*p->vx, p->vy);
                pressure += p->mass * 2 * p->vx / height;
            }
            if ( (p->y - p->radius <= 0) and (p->vy < 0)){
                p->setVelocities(p->vx, -1*p->vy);
                pressure += p->mass * 2 * p->vy / width;
            }
        }
        pressureSteps.push_back(pressure);


        // next check all pairwise combos of particles
        // for collisions, update velocities if necessary
        // upper triangular nested loop
        for (int ii=0; ii < nParticles - 1; ii++){
            for (int jj=ii+1; jj < nParticles; jj++){
                auto p1 = particles[ii];
                auto p2 = particles[jj];
                if (isCollided(p1, p2)){
                    handleCollision(p1, p2);
                    collisions += 1;
                }
            }
        }
        collisionSteps.push_back(collisions);


        if ((step % saveEvery) == 0){
            double fraction = float(step)/(float)steps;
            std::cout << "saving at step " << step << " fraction done " << fraction << std::endl;
            particleSteps.push_back(dumpState());
        }

    }

}







